conda/core/python_dist.py

```diff
@@ -0,0 +1,444 @@
+    'Requires-Dist',
+    'Requires-External',
+    'Project-URL',
+    'Provides-Extra',
+    'Provides-Dist',
+    'Obsoletes-Dist',
+)
+
+
+class MetadataConflictError(Exception):
+    pass
+
+
+class PythonDistributionMetadata(object):
+    """
+    The canonical method to transform metadata fields into such a data
+    structure is as follows:
+      - The original key-value format should be read with
+        email.parser.HeaderParser
+      - All transformed keys should be reduced to lower case. Hyphens should
+        be replaced with underscores, but otherwise should retain all other
+        characters
+      - The transformed value for any field marked with "(Multiple-use")
+        should be a single list containing all the original values for the
+        given key
+      - The Keywords field should be converted to a list by splitting the
+        original value on whitespace characters
+      - The message body, if present, should be set to the value of the
+        description key.
+      - The result should be stored as a string-keyed dictionary.
+
+    Notes
+    -----
+      - Metadata 2.1: https://www.python.org/dev/peps/pep-0566/
+      - Metadata 2.0: https://www.python.org/dev/peps/pep-0426/
+      - Metadata 1.2: https://www.python.org/dev/peps/pep-0345/
+      - Metadata 1.1: https://www.python.org/dev/peps/pep-0314/
+      - Metadata 1.0: https://www.python.org/dev/peps/pep-0241/
+      - https://packaging.python.org/specifications/core-metadata/
+    """
+
+    # TODO: Define precedence
+    METADATA_FILE_NAMES = ('METADATA', 'PKG-INFO', 'metadata.json')
+
+    def __init__(self, path):
+        """"""
+        self._path = path
+
+        metadata_path = self._process_metadata_path(path)
+        self._read_metadata(metadata_path)
+
+    def _process_metadata_path(self, path):
+        """"""
+        if isdir(path):
+            for fname in self.METADATA_FILE_NAMES:
+                fpath = join(path, fname)
+                if isfile(fpath):
+                    metadata_path = fpath
+                    break
+        elif isfile(path):
+            # '<pkg>.egg-info' file contains metadata directly
+            metadata_path = path
+        else:
+            raise MetadataConflictError
+
+        return metadata_path
+
+    def _read_metadata(self, fpath):
+        """"""
+        if isfile(fpath):
+            if fpath.endswith('.json'):
+                self._data = self._read_json_metadata(fpath)
+            else:
+                self._data = self._parse_metadata(fpath)
+        else:
+            self._data = {}
+
+    def _read_json_metadata(self, fpath):
+        """"""
+        data = None
+        if isfile(fpath):
+            with open(fpath, 'r') as f:
+                data = json.loads(f.read())
+        return data
+
+    def _parsed_metadata_to_json(self, data):
+        """Parse the original format which is stored as RFC-822 headers."""
+        new_data = OrderedDict() if data else None
```

Comment:
Use `conda.common.compat.odict()` instead of `OrderedDict()`.

Reply:
Why `else None`?  Are you wanting to force a NoneType error somewhere?

Reply:
Well... if there is no data (empty metadata we should fail somehow.) so maybe adding an actual exception @kalefranz ?